### ISSUE TITLE
Custom operators

### ISSUE THREAD

**Issue by ivan-kleshnin** [caolan/highland#269]
RxJS [allows](https://github.com/ReactiveX/RxJava/wiki/Implementing-Your-Own-Operators) to implement your own custom operators. 

Working with complex data like Vinyl objects in Gulp streams it would be great to have such option.

Here how it loooks like now, a lot of boilerplate and repeating code:

```javascript
.map(file => {
  let json = file.contents.toString();
  let js = JSON.parse(json);
  // replace one field in JSON
  let str = ...
  file.contents = new Buffer(str);
  return file;
})
```

It may be like this:
```javascript
.json.map(value => {
  // produce newValue
  return newValue;
})
```

So `json.map` just handles all this mess and your Gulp pipes should look much more clear.

**Comment by vqvu** [comment]
```javascript
function asObject(xs) {
    // Do whatever you want to the stream here.
    // return a new stream. Typically uses `consume`.
}

_([]).through(asObject)
    ...
```

This is equivalent to RxJava's `transform`(https://github.com/ReactiveX/RxJava/wiki/Implementing-Your-Own-Operators#transformational-operators). Writing the actual operator can be done using `consume` and `pull` (all core highland operators are implemented like this), but can also just compose a bunch of pre-existing operators.

**Comment by ivan-kleshnin** [comment]
@vqvu thanks for the answer

> _([]).through(asObject)

Well, I may also use through2 or low-level stream operations. But I want to make the operators, not functions to not bother with imports into tens of files in my case.

> Writing the actual operator can be done using consume and pull (all core highland operators are implemented like this), but can also just compose a bunch of pre-existing operators.

Again, I may inspect Highland source or monkeypatch, I junst wanted to request example / official documentation for a future.

**Comment by vqvu** [comment]
Oh, you mean patch the prototype directly to add methods (the "operator" wording confused me)? There's no official way to do that. You pretty much have to do gulp-syle `.through(something())` if you want really custom stuff.

You can use `Object.getPrototypeOf`, but again, not officially supported.

If you're worried about importing the same function everywhere, maybe you can structure your code so that the `asObject` and `asVinyl` happens automatically in a single place? Personally, I don't like patching prototypes, because it introduces a hidden dependency on the patching code that can't be seen just from reading the source.

**Comment by ivan-kleshnin** [comment]
@vqvu yeah I speak in RxJS terminology, familiar to me. I think there is a difference between library and application. I agree that for library case custom operators really have drawbacks. But for application code where everything is under my control potential drawbacks are very unprobable and benefits are obvious. That's why RxJS still provides some functionality to add custom ops IMO.

**Comment by vqvu** [comment]
Oh, what I meant was that a function used with `through` is still an operator, same as something in the prototype.

There's some discussion on why we don't expose the prototype in #148. It mostly has to do with having multiple versions of highland at once (when using NPM).

In any case, I do agree that application code can do things that library code doesn't, since you control application code directly (still don't like patching the prototype but I don't want to get into a coding style discussion here). So for your use case, you should be able to go ahead and do

```javascript
var prototype = Object.getPrototypeOf(_());
prototype.myOperator = function (...) {
...
}
```

This will work with the current version of library (and probably with any future 2.x version).

**Comment by ivan-kleshnin** [comment]
There was also a similar talk at https://github.com/caolan/highland/issues/70
Ok, I'm gonna try all this and get some real-world impression. Node streams are way complicated unfortunately.

**Comment by vqvu** [comment]
Yep. Among others!

If you come up with a good way to do this, please share.

**Comment by ivan-kleshnin** [comment]
Custom operators could be easily added if we had functional API instead of OOP.

```
import {pipe} from "ramda"

...
pipe(
  HL.map(...), // curryable function, data is last argument
  HL.filter(...)  // curryable function, data is last argument
 )(stream)
```

No namespacing issues because no rigid artificial namespacing OOP enforces.

**Comment by vqvu** [comment]
We have this. All transforms have a top-level curried version, and the only way to implement custom operators is by using a functional, data-last, style. Whether you call them in a functional style

```javascript
pipe(
    customFn1(...),
    customFn2(...)
)(stream)
```

or a more OO style

```javascript
stream.through(customFn1(...))
    .through(customFn2(...))
```

is simply a matter of preference. It has nothing to do with how easy or hard it is to add custom operators.

I mean, Highland doesn't have the `pipe` function, but that's easily implemented (by yourself or gotten from ramda).

**Comment by ivan-kleshnin** [comment]
> is simply a matter of preference. It has nothing to do with how easy or hard it is to add custom operators.

What? With OO style you need to inject function into object (library code).
With Functional style you just need to import function before call (app code).

**Comment by vqvu** [comment]
> With OO style you need to inject method into object (library code) to be able to use it.

I consider using `through` to be OO style, and `through` doesn't require you to inject any thing.

I really don't understand what your concern is. Right this moment, you can choose between using OO or functional style.

For core operators, you can use a functional style

```javascript
pipe(
    _.map(mapper),
    _.filter(f)
)(stream);
```
or an OO style
```javascript
stream.map(mapper)
    .filter(f);
```

For custom function, you can do the exact same thing, as I've already mentioned.

The way you implement custom operators is the same, and you can choose between OO or functional style. How can you say it's not a matter of preference? We are talking about how *custom operators* can be implemented, not how the core library is implemented.

> This namespacing issues is direct consequence of OOP style.

Actually, it's a direct consequence of NPM allowing multiple versions of the same library to coexist in a single program. It'd be worse if the core library was implemented purely functionally, with data and operators separate. You'd end up with different data format for different versions of highland, which requires different versions of operators (or at least the `consume` and `pull` methods). You'd have no way of knowing which version to use for a particular stream. The OO style at least helps you select the right version and gives you a common API (i.e., `consume`) to code custom operators against.

But I don't see how it's relevant to the issue at hand. Again, it's not like you *can't* use the functional style for custom operators.

**Comment by ivan-kleshnin** [comment]
> I really don't understand what your concern is. Right this moment, you can choose between using OO or functional style.

Ok, I didn't get you support both styles. My mistake. 

> Actually, it's a direct consequence of NPM allowing multiple versions of the same library to coexist in a single program.

Did you meant "not allowing"? Because you can't 
```js
import HighLand2_5_1 from "highland@2.5.1";
import HighLand2_5_2 from "highland@2.5.2";
```

**Comment by vqvu** [comment]
You can't directly depend on two different versions of a library. But if you depend on some library, that *that* library depends on highland, you can run into a situation where your dependency depends on highland@2.1.0 while you have highland@2.5.0 as a separate dependency. The issue arises when that library leaks its dependency on Highland. For example, if that library returns a Highland stream, it'll be a stream constructed with highland@2.1.0 and not your preferred highland@2.5.0.

**Comment by ivan-kleshnin** [comment]
Thank you. It is really a separate issue and a very tricky one. I can't help with that unfortunately. 
Not enough competency.
